src/DI/ContainerLoader.php

```diff
@@ -61,46 +61,57 @@ public function getClassName($key)
 	private function loadFile($class, $generator)
 	{
 		$file = "$this->tempDirectory/$class.php";
-		if (!$this->autoRebuild && (@include $file) !== FALSE) { // @ - file may not exist
+		if (!$this->isExpired($file) && (@include $file) !== FALSE) {
 			return;
 		}
 
 		if (!is_dir($this->tempDirectory)) {
 			@mkdir($this->tempDirectory); // @ - directory may already exist
 		}
+
 		$handle = fopen("$file.lock", 'c+');
-		if (!$handle) {
-			throw new Nette\IOException("Unable to open or create file '$file.lock'.");
+		if (!$handle || !flock($handle, LOCK_EX)) {
+			throw new Nette\IOException("Unable to acquire exclusive lock '$file.lock'.");
 		}
 
-		if ($this->autoRebuild) {
-			flock($handle, LOCK_SH);
-			foreach ((array) @unserialize(file_get_contents("$file.meta")) as $f => $time) { // @ - file may not exist
-				if (@filemtime($f) !== $time) { // @ - stat may fail
-					@unlink($file); // @ - file may not exist
-					break;
-				}
+		if (!is_file($file) || $this->isExpired($file)) {
+			list($code, $dependencies) = call_user_func($generator, $class);
+
+			$dependencies = (array) $dependencies;
+			$meta = serialize(array_combine($dependencies, @array_map('filemtime', $dependencies)));
+			if (file_put_contents("$file.tmp", $meta) !== strlen($meta) || !rename("$file.tmp", "$file.meta")) {
+				@unlink("$file.tmp"); // @ - file may not exist
+				throw new Nette\IOException("Unable to create file '$file.meta'.");
+			}
+
+			$code = "<?php\n" . $code;
+			if (file_put_contents("$file.tmp", $code) !== strlen($code) || !rename("$file.tmp", $file)) {
```

Comment:
What about something like

`if (file_put_contents("$file.tmp1", $meta) !== strlen($meta) || file_put_contents("$file.tmp2", $code) !== strlen($code) 
|| !rename("$file.tmp1", "$file.meta") || !rename("$file.tmp", $file))`

ie. first write both files and then rename them.